### PR TITLE
Add a job that will build documentation and create PR

### LINKED ISSUE #651: Generate and publish documentation on CI
We should adapt the work @chezzdev did in mapbox/mapbox-navigation-ios#3439 mapbox/mapbox-navigation-ios#3469 to run the jazzy documentation generation script on CI for every PR and push it to publisher-production for release tags. This will completely automate two of the steps of [the release process](https://github.com/mapbox/mapbox-directions-swift/blob/main/CONTRIBUTING.md#releasing-a-new-version). It would also obviate #650, since no one would need to run publish-documentation.sh manually anymore.

/cc @mapbox/navigation-ios @mapbox/docs

### PR DESCRIPTION
**Pull request by S2Ler**
I'm no longer able to run jazzy locally and decided to move the task to a stable environment. 

Fixes #651, although #651 says that docs should be published for each PR. I don't think it is necessary. But can be solved separately.

- [x] Update [Contributing Guide](https://github.com/mapbox/mapbox-directions-swift/blob/main/CONTRIBUTING.md) after general idea is accepted.
- [x] Test this job e2e.

### CODE REVIEW

File: .circleci/config.yml
Comment (bamx23) [inline .circleci/config.yml:219]: I would recommend splitting this step into several separate ones:
* Build Documentation
* Setup Repo Write Access
* Create Docs PR

File: .circleci/config.yml
Comment (bamx23) [inline .circleci/config.yml:225]: By the way, do we need a PR? Why dot pushing directly to `publisher-production` as we do in Navigation SDK repo?

Comment (bamx23) [inline .circleci/config.yml:225]: Does this command create a PR to `publisher-production` branch? And I'm not sure that reviewer group will be added (there was an issue with this for `mbx-ci` generated tokens)

File: .circleci/config.yml
Comment (bamx23) [inline .circleci/config.yml:263]: We can run this for all valid version tags (without approval)

File: scripts/publish-documentation.sh
Comment (bamx23) [inline scripts/publish-documentation.sh:38]: Should we remove `--depth=1 --prune` from `git fetch --depth=1 --prune` above _(line 17)_?

File: .circleci/config.yml
Comment (bamx23) [inline .circleci/config.yml:257]: Just as an idea: 
* We can build docs for every commit and store it [in artifacts](https://github.com/mapbox/mapbox-navigation-ios/blob/fc37d6661e7ef73959e0a30108f00b0e487e281b/.circleci/config.yml#L648-L650)
* Then there is an approval job for publishing (that works is for tags only)
* And once approved we can push docs to `publisher-production`

File: scripts/publish-documentation.sh
Comment (1ec5) [reply scripts/publish-documentation.sh:38]: Yes, please remove the shallow clone. I always avoided the publish-documentation.sh script because this step trashes my clone.

**Review by Udumft — APPROVED** [review]
LGTM

**Pull request merged**
